src/ProjectTemplates/Web.ProjectTemplates/content/BlazorServerWeb-CSharp/Areas/Identity/RevalidatingAuthenticationStateProvider.cs

```diff
@@ -1,98 +1,59 @@
         {
-            var cancellationToken = _loopCancellationTokenSource.Token;
-
-            while (!cancellationToken.IsCancellationRequested)
+            // Get the user manager from a new scope to ensure it fetches fresh data
+            using (var scope = _scopeFactory.CreateScope())
             {
-                try
-                {
-                    await Task.Delay(RevalidationInterval, cancellationToken);
-                }
-                catch (TaskCanceledException)
-                {
-                    break;
-                }
-
-                var isValid = await CheckIfAuthenticationStateIsValidAsync();
-                if (!isValid)
-                {
-                    // Force sign-out. Also stop the revalidation loop, because the user can
-                    // only sign back in by starting a new connection.
-                    var anonymousUser = new ClaimsPrincipal(new ClaimsIdentity());
-                    _currentAuthenticationStateTask = Task.FromResult(new AuthenticationState(anonymousUser));
-                    NotifyAuthenticationStateChanged(_currentAuthenticationStateTask);
-                    _loopCancellationTokenSource.Cancel();
-                }
+                var userManager = scope.ServiceProvider.GetRequiredService<UserManager<TUser>>();
+                return await ValidateSecurityStampAsync(userManager, authenticationState.User);
             }
         }
 
-        private async Task<bool> CheckIfAuthenticationStateIsValidAsync()
+        private async Task<bool> ValidateSecurityStampAsync(UserManager<TUser> userManager, ClaimsPrincipal principal)
         {
-            try
+            var user = await userManager.GetUserAsync(principal);
+            if (user == null)
             {
-                // Get the sign-in manager from a new scope to ensure it fetches fresh data
-                using (var scope = _scopeFactory.CreateScope())
-                {
-                    var signInManager = scope.ServiceProvider.GetRequiredService<SignInManager<TUser>>();
-                    var authenticationState = await _currentAuthenticationStateTask;
-                    var validatedUser = await signInManager.ValidateSecurityStampAsync(authenticationState.User);
-                    return validatedUser != null;
-                }
+                return false;
             }
-            catch (Exception ex)
+            else if (!userManager.SupportsUserSecurityStamp)
             {
-                _logger.LogError(ex, "An error occurred while revalidating authentication state");
-                return false;
+                return true;
+            }
+            else
+            {
+                var principalStamp = principal.FindFirstValue(_options.ClaimsIdentity.SecurityStampClaimType);
+                var userStamp = await userManager.GetSecurityStampAsync(user);
+                return principalStamp == userStamp;
```

Comment:
@HaoK

Reply:
Looks good, and yeah much more readable than the SignInManager code